File: tooling/php/contracts.php
<?php
declare(strict_types=1);

namespace Refresher;

interface KeyValueStore
{
    public function put(string $key, string $value, int $ttlMs): void;
    public function get(string $key): ?string;
}

interface JobQueuePort
{
    public function enqueue(string $payload): string;
    /** @return array{job_id:string,payload:string,attempt:int,receipt:string}|null */
    public function reserve(): ?array;
    public function ack(string $receipt): bool;
    public function nack(string $receipt): bool;
}

File: tooling/python/contracts.py
"""Structural ports for the dispatcher; these contain no implementations."""
from collections.abc import Callable
from typing import Protocol, TypedDict

Clock = Callable[[], int]
Sender = Callable[[str], bool]


class Delivery(TypedDict):
    job_id: str
    payload: str
    attempt: int
    receipt: str


class KeyValueStore(Protocol):
    def put(self, key: str, value: str, ttl_ms: int) -> None: ...
    def get(self, key: str) -> str | None: ...


class JobQueuePort(Protocol):
    def enqueue(self, payload: str) -> str: ...
    def reserve(self) -> Delivery | None: ...
    def ack(self, receipt: str) -> bool: ...
    def nack(self, receipt: str) -> bool: ...
